src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -0,0 +1,212 @@
+
+#include "ray/core_worker/transport/direct_actor_transport.h"
+#include "ray/common/task/task.h"
+
+using ray::rpc::ActorTableData;
+
+namespace ray {
+
+bool HasByReferenceArgs(const TaskSpecification &spec) {
+  for (int i = 0; i < spec.NumArgs(); ++i) {
+    if (spec.ArgIdCount(i) > 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+CoreWorkerDirectActorTaskSubmitter::CoreWorkerDirectActorTaskSubmitter(
+    boost::asio::io_service &io_service, gcs::RedisGcsClient &gcs_client,
+    CoreWorkerObjectInterface &object_interface)
+    : io_service_(io_service),
+      gcs_client_(gcs_client),
+      client_call_manager_(io_service),
+      store_provider_(
+          object_interface.CreateStoreProvider(StoreProviderType::LOCAL_PLASMA)) {
+  RAY_CHECK_OK(SubscribeActorUpdates());
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::SubmitTask(
+    const TaskSpecification &task_spec) {
+  if (HasByReferenceArgs(task_spec)) {
+    return Status::Invalid("direct actor call only supports by-value arguments");
+  }
+
+  RAY_CHECK(task_spec.IsActorTask());
+  const auto &actor_id = task_spec.ActorId();
+
+  const auto task_id = task_spec.TaskId();
+  const auto num_returns = task_spec.NumReturns();
+
+  auto request = std::unique_ptr<rpc::PushTaskRequest>(new rpc::PushTaskRequest);
+  request->set_task_id(task_spec.TaskId().Binary());
+  request->mutable_task_spec()->CopyFrom(task_spec.GetMessage());
+
+  std::unique_lock<std::mutex> guard(rpc_clients_mutex_);
+  auto iter = actor_states_.find(actor_id);
+  if (iter == actor_states_.end() ||
+      iter->second.state_ == ActorTableData::RECONSTRUCTING) {
+    // Actor is not yet created, or is being reconstructed, cache the request
+    // and submit after actor is alive.
+    // TODO(zhijunfu): it might be possible for a user to specify an invalid
+    // actor handle (e.g. from unpickling), in that case it might be desirable
+    // to have a timeout to mark it as invalid if it doesn't show up in the
+    // specified time.
+    auto pending_request = std::unique_ptr<PendingTaskRequest>(
+        new PendingTaskRequest(task_id, num_returns, std::move(request)));
+    pending_requests_[actor_id].emplace_back(std::move(pending_request));
+    return Status::OK();
+  } else if (iter->second.state_ == ActorTableData::ALIVE) {
+    // Actor is alive, submit the request.
+    if (rpc_clients_.count(actor_id) == 0) {
+      // If rpc client is not available, then create it.
+      HandleActorAlive(actor_id, iter->second.location_.first,
```

Comment:
The name `HandleActorAlive` isn't proper here, maybe rename to `CreateConnectionAndSendPendingTasks`?

Reply:
yep, it's just that this seems a bit too long... 
How about `ConnectAndSendPendingTasks` ?

Reply:
sounds good